### PR TITLE
Client IPC Interface

### PR DESCRIPTION
**Pull request by NikVolf**
IPC interface for Client

from example test to run client:

```
	let remote_client = nanoipc::init_client::<RemoteClient<_>>(socket_path).unwrap();
	assert!(remote_client.handshake().is_ok());
```

from example test to run worker:

```
	let mut worker = nanoipc::Worker::new(&client);
	worker.add_reqrep(&socket_path).unwrap();
	while !stop.load(Ordering::Relaxed) {
		worker.poll();
	}
```

Progress tracker

- [x] limiting errors to crate-level error types #1525
- [x] void rpc methods missing #1540
- [x] interface tweaks mapping
- [x] some docs missing
- [x] pretty name configuration #1500 
- [x] simple rpc test

### CODE REVIEW

**Comment by gavofyork** [comment]
lgtm, but should get a few other eyes on it, too...

**Comment by arkpar** [comment]
Can we keep client code in the `rs` file instead if `rs.in`?
Why some types (e.g. `BlockQueueInfo`) are moved to different modules? `BlockQueueInfo` belongs with the `BlockQueue`

**Comment by NikVolf** [comment]
@arkpar 
yeah why not just `.rs`
i thought it was some kind of convention for `rs.in`, but can't see any proofs now

`BlockQueueInfo` moved to `src/types` because this is the only directory subject to serialization codegen pass

and this kind of entries should be kept minimum until compiler plugins stabilized, because you will get errors in generated code rather than original one

so until then everything that is subject to serialization, lives in `src/types/*`

**Comment by NikVolf** [comment]
@ark `.rs.in` -> `.rs`

**Comment by NikVolf** [comment]
@gavofyork merged with master

**Comment by rphmeier** [comment]
still says there are merge conflicts

**Comment by arkpar** [comment]
The file containing most of the clients code is called `client_ipc.rs`?
We do still want to have no-ipc build, right?
Client code should be ipc-independent ideally

**Comment by NikVolf** [comment]
@rphmeier 
because i have to merge conflicts every time some commits to the client.rs :)

@arkpar 
everything is completely independent, if you create just `Client`, you work without any ipc
if you create `RemoteClient`, you work via ipc

**Comment by gavofyork** [comment]
i have to agree with @arkpar - why is `Client` class in a file called `client_ipc.rs`? unless there's good reasons, filenames should reflect the main class in them.

**Comment by NikVolf** [comment]
@arkpar @gavofyork 
right, didn't got the concerns right
actually extra file can be avoided completely
and diff if much cleaner now

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -811,18 +837,17 @@ impl BlockChainClient for Client {
...
 								block_hash: hash.clone(),
 								block_number: number,
 								transaction_hash: hashes.get(index).cloned().unwrap_or_else(H256::new),
 								transaction_index: index,
 								log_index: log_index + i
-							})
-							.collect::<Vec<LocalizedLogEntry>>()
-					})
+						})
 					.collect::<Vec<LocalizedLogEntry>>()
```
Comment (gavofyork) [inline ethcore/src/client/client.rs]: indentation messed up.

**Comment by gavofyork** [comment]
looks good apart from the errant indent.

**Comment by NikVolf** [comment]
@gavofyork fixed

**Pull request merged**